<a href="https://colab.research.google.com/github/alfa7g7/Analitica-prescriptiva/blob/main/Talleres/Taller_1_Analitica_Prescriptiva.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TALLER 1 | ANALITICA PRESCRIPTIVA | EXCEL SOLVER - PYTHON PULP

### INTEGRANTES:
              Arlex Pino
              Raul Echeverry
              Esteban Ordoñez
              Fabian Salazar Figueroa



In [ ]:
#librerias a usar
!pip install pulp #para usar analitica prescriptiva 1

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.7/17.7 MB 20.4 MB/s eta 0:00:00


##**1. FACTORIES ON THE MOMISS RIVER**
There are three factories on the Momiss River (1, 2, and 3).

Each emits two types of pollutants (1 and 2) into the river. If the waste from each factory is processed, the pollution in the river can be reduced.
  
It costs $15 to process a ton of factory 1 waste, and each ton processed reduces the amount of pollutant 1 by 0.10 ton and the amount of pollutant 2 by 0.45 ton.

It costs $10 to process a ton of factory 2 waste, and each ton processed will reduce the amount of pollutant 1 by 0.20 ton and the amount of pollutant 2 by 0.25 ton.

It costs $20 to process a ton of factory 3 waste, and each ton processed will reduce the amount of pollutant 1 by 0.40 ton and the amount of pollutant 2 by 0.30 ton.

The state wants to reduce the amount of pollutant 1 in the river by at least 30 tons and the amount of pollutant 2 in the river by at least 40 tons.

Formulate an LP that will minimize the cost of reducing pollution by the desired amounts. Do you think that the LP assumptions (Proportionality, Additivity, Divisibility, and Certainty) are reasonable for this problem?

In [ ]:
# Importamos la librería
from pulp import *

# Definición de conjuntos
fabricas = [1, 2, 3]  # Conjunto de fábricas
contaminantes = [1, 2]  # Tipos de contaminantes

# Definición de parámetros
costos = {1: 15, 2: 10, 3: 20}  # Costo de tratamiento por fábrica

reduccion = {
    1: {1: 0.1, 2: 0.45},  # Fábrica 1 reduce 0.1 de contaminante 1 y 0.45 de contaminante 2 por tonelada
    2: {1: 0.2, 2: 0.25},  # Fábrica 2 reduce 0.2 de contaminante 1 y 0.25 de contaminante 2 por tonelada
    3: {1: 0.4, 2: 0.3}    # Fábrica 3 reduce 0.4 de contaminante 1 y 0.3 de contaminante 2 por tonelada
}

min_reduccion = {1: 30, 2: 40}  # Requerimientos mínimos de reducción

# Definición de instancia del problema
prob = LpProblem("Minimizar_Costo_Tratamiento", LpMinimize)

# Definición de variables de decisión
X = LpVariable.dicts(name="X", indices=fabricas, lowBound=0, upBound=None, cat=LpContinuous,)  # Cantidad de toneladas tratadas en cada fábrica

# Función objetivo: minimizar el costo total de tratamiento
prob += lpSum(costos[f] * X[f] for f in fabricas), "Costo_Total"

# Restricciones de reducción mínima de contaminantes
for p in contaminantes:
    prob += lpSum(reduccion[f][p] * X[f] for f in fabricas) >= min_reduccion[p], f"Reducción_Contaminante_{p}"

# Resolver el problema
prob.solve()

# Imprimir resultados
for f in fabricas:
    print(f"La fábrica {f} debe procesar una cantidad X{f} =", X[f].varValue," toneladas de desecho")

print("Costo total mínimo =", prob.objective.value())


La fábrica 1 debe procesar una cantidad X1 = 7.6923077  toneladas de desecho
La fábrica 2 debe procesar una cantidad X2 = 146.15385  toneladas de desecho
La fábrica 3 debe procesar una cantidad X3 = 0.0  toneladas de desecho
Costo total mínimo = 1576.9231155000002


### **Formulación del problema y evaluación de los supuestos de Programación Lineal**  

Para minimizar el costo de reducción de la contaminación hasta los niveles requeridos, se ha formulado un modelo de Programación Lineal (LP) donde la función objetivo es minimizar el costo total de procesamiento de los desechos de las fábricas, sujeto a restricciones que garantizan la reducción mínima de contaminantes.  

En cuanto a los **supuestos de Programación Lineal**, evaluamos si son razonables en este problema:  

1. **Proporcionalidad** ✅  
   El modelo asume que la cantidad de contaminantes reducidos es directamente proporcional a la cantidad de desechos tratados en cada fábrica. Dado que cada tonelada procesada reduce una cantidad fija de contaminantes, este supuesto es válido.  

2. **Aditividad** ✅  
   La contribución de cada fábrica a la reducción total de contaminantes se suma linealmente. No hay interacciones complejas entre las fábricas que alteren la efectividad de la reducción, por lo que este supuesto también se cumple.  

3. **Divisibilidad** ✅  
   Se permite que la cantidad de desechos procesados en cada fábrica sea un número real (fraccionario). En la práctica, los tratamientos pueden aplicarse de manera continua y no solo en cantidades enteras, por lo que la divisibilidad es un supuesto razonable.  

4. **Certeza** ⚠️  
   Aquí el supuesto puede ser debatible. El modelo asume que los costos y las tasas de reducción de contaminación son valores fijos y conocidos con certeza. Sin embargo, en la realidad, podrían variar debido a factores externos como fluctuaciones en costos operativos, eficiencia del procesamiento o condiciones ambientales.  

### **Conclusión**  
En general, los supuestos de **Proporcionalidad, Aditividad y Divisibilidad** son razonables y apropiados para este problema. Sin embargo, el supuesto de **Certeza** podría no cumplirse del todo en un entorno real, donde los costos y la efectividad de la reducción pueden verse afectados por diversas variables. En un análisis más avanzado, se podría considerar una formulación más robusta que tenga en cuenta incertidumbre en estos parámetros.

### 📌 Solución en Excel  
Este punto puedes verlo resuelto en Excel a través de Solver en el siguiente enlace en la hoja llamada "Ejercicio 1":  
[Haz clic aquí para verlo](https://docs.google.com/spreadsheets/d/1aVTQIgdwYh91xQ1oS19cvkOjbDoZUB89/edit?usp=sharing&ouid=102065898380876538162&rtpof=true&sd=true)

##**2. POST OFFICE SITUATIONS**


###**A. the post office full-time employees who must be hired.**
A post office requires different numbers of full-time employees on different days of the week.

The number of full-time employees required on each day is given in Table 4.

Union rules state that each full-time employee must work five consecutive days and then receive two days off.

For example, an employee who works Monday to Friday must be off on Saturday and Sunday.

The post office wants to meet its daily requirements using only fulltime employees.

Formulate an LP that the post office can use to minimize the number of full-time employees who must be hired.


In [ ]:
# Definición de conjuntos
# Definir los días de la semana
dias = ["Lunes", "Martes", "Miércoles", "Jueves", "Viernes", "Sábado", "Domingo"]

# Definición de parámetros
# Demanda diaria de empleados
requeridos = {"Lunes": 17, "Martes": 13, "Miércoles": 15, "Jueves": 19, "Viernes": 14, "Sábado": 16, "Domingo": 11}

# Definición de instancia del problema
# Crear problema de minimización
prob = LpProblem("Minimizar_Empleados", LpMinimize)

# Definición de variables de decisión
# Variables de decisión: empleados que comienzan en cada día
X = LpVariable.dicts("X", dias, 0, None, LpInteger)

# Función objetivo: minimizar el número total de empleados contratados
prob += lpSum(X[d] for d in dias)

# Restricciones: cubrir la demanda de cada día considerando 5 días consecutivos de trabajo
for i, d in enumerate(dias):
    prob += (X[dias[i % 7]] + X[dias[(i - 1) % 7]] + X[dias[(i - 2) % 7]] +
             X[dias[(i - 3) % 7]] + X[dias[(i - 4) % 7]] >= requeridos[d])

# Resolver el problema
prob.solve()

# Imprimir resultados
for d in dias:
    print(f"Empleados que comienzan el {d}: {X[d].varValue}")

print("Número mínimo total de empleados contratados:", value(prob.objective))


Empleados que comienzan el Lunes: 6.0
Empleados que comienzan el Martes: 3.0
Empleados que comienzan el Miércoles: 2.0
Empleados que comienzan el Jueves: 8.0
Empleados que comienzan el Viernes: 0.0
Empleados que comienzan el Sábado: 3.0
Empleados que comienzan el Domingo: 1.0
Número mínimo total de empleados contratados: 23.0


#### **Conclusión y Recomendaciones para la Asignación de Empleados (Punto A)**  

El modelo de optimización ha determinado la **cantidad mínima de empleados** necesarios para cubrir la demanda laboral de la semana, asignándolos estratégicamente a lo largo de los días.  



#### **Distribución Óptima de Contrataciones**  

| Día de Inicio | Número de Empleados Contratados |
|--------------|--------------------------------|
| **Lunes**    | 6  |
| **Martes**   | 3  |
| **Miércoles** | 2  |
| **Jueves**   | 8  |
| **Viernes**   | 0  |
| **Sábado**   | 3  |
| **Domingo**  | 1  |

✅ **Número total mínimo de empleados contratados:** **23**  



#### **Puntos Clave y Recomendaciones**  

1. **Distribución eficiente del personal**  
   - El modelo encontró que **23 empleados son suficientes** para cubrir la demanda laboral, minimizando costos.  
   - La mayoría de los empleados comienzan su trabajo los jueves (8 personas), lo que sugiere que hacia el final de la semana hay mayor necesidad de mano de obra.  

2. **No se requiere contratación los viernes**  
   - El modelo indica que no es necesario contratar empleados nuevos los viernes, lo que puede deberse a que la carga laboral es menor en ese punto de la semana.  

3. **El jueves es el día con más contrataciones**  
   - Con **8 empleados iniciando el jueves**, se evidencia que es un día de alta necesidad de personal.  
   - Se podría analizar si es posible redistribuir parte de esta carga laboral para equilibrar mejor la semana.  

4. **Optimización en la asignación del personal**  
   - Se redujo al **mínimo posible** la cantidad de empleados sin afectar la operación.  
   - Esto implica una **buena planificación de horarios y turnos** para maximizar la productividad con menos contrataciones.  


#### **Recomendación Final**  
La **Post Office** debería seguir este plan de contratación para operar con el **mínimo número de empleados posible (23),** asegurando que los días clave con mayor demanda (especialmente jueves y lunes) estén cubiertos de manera óptima.  

También sería útil revisar la carga de trabajo para evaluar si **redistribuir empleados** entre los días puede mejorar la eficiencia aún más.  

**Conclusión:** Se logró minimizar la contratación de empleados sin afectar la operación, garantizando **eficiencia y rentabilidad** en la asignación de personal.

### 📌 Solución en Excel  
Este punto puedes verlo resuelto en Excel a través de Solver en el siguiente enlace en la hoja llamada "Ejercicio 2 A.":  
[Haz clic aquí para verlo](https://docs.google.com/spreadsheets/d/1aVTQIgdwYh91xQ1oS19cvkOjbDoZUB89/edit?usp=sharing&ouid=102065898380876538162&rtpof=true&sd=true)

###**B. The post office’s weekly labor costs**
In the post office example, suppose that each full-time employee works 8 hours per day.

Thus, Monday’s requirement of 17 workers may be viewed as a requirement of 8(17) 136 hours.

The post office may meet its daily labor requirements by using both full-time and part-time employees.

During each week, a full-time employee works 8 hours a day for five consecutive days, and a part-time employee works 4 hours a day for five consecutive days.

A full-time employee costs the post office 15 USD per hour, where as a part-time employee (with reduced fringe benefits) costs the post office only 10 USD per hour.

Union requirements limit part-time labor to 25% of weekly labor requirements.

Formulate an LP to minimize the post office’s weekly labor costs.


In [ ]:
# Forma A
# Definir los días de la semana
days = ["Lunes", "Martes", "Miercoles", "Jueves", "Viernes", "Sábado", "Domingo"]

# Demanda de empleados por día (en horas)
demand = {"Lunes": 136, "Martes": 104, "Miercoles": 120, "Jueves": 152, "Viernes": 112, "Sábado": 128, "Domingo": 88}

# Definir variables de decisión
X_FT = LpVariable.dicts("FT", days, 0, None, cat="Integer") # Empleados full-time
X_PT = LpVariable.dicts("PT", days, 0, None, cat="Integer") # Empleados part-time

# Definir el problema de optimización
prob = LpProblem("Minimizar_Costo_Empleados", LpMinimize)

# Función objetivo corregida: Minimizar costos semanales reales
prob += lpSum([15 * 8 * 5 * X_FT[d] + 10 * 4 * 5 * X_PT[d] for d in days])

# Restricciones de horas requeridas por día
for i, d in enumerate(days):
    prob += (lpSum(8 * X_FT[days[j % 7]] + 4 * X_PT[days[j % 7]] for j in range(i, i + 5)) >= demand[d])

# Restricción del 25% de empleados medio tiempo (en términos de horas semanales)
total_hours_week = sum(demand.values())
prob += lpSum(4 * 5 * X_PT[d] for d in days) <= 0.25 * total_hours_week

# Resolver el problema
prob.solve()

# Imprimir resultados
for d in days:
    print(f"Día {d}: FT = {X_FT[d].varValue}, PT = {X_PT[d].varValue}")
print("Costo total mínimo =", prob.objective.value())


Día Lunes: FT = 7.0, PT = 1.0
Día Martes: FT = 0.0, PT = 0.0
Día Miercoles: FT = 3.0, PT = 0.0
Día Jueves: FT = 0.0, PT = 0.0
Día Viernes: FT = 3.0, PT = 7.0
Día Sábado: FT = 4.0, PT = 1.0
Día Domingo: FT = 1.0, PT = 0.0
Costo total mínimo = 12600.0


In [ ]:
#Forma B
# Definir los días de la semana
days = ["Lunes", "Martes", "Miércoles", "Jueves", "Viernes", "Sábado", "Domingo"]
num_days = len(days)

# Demanda de empleados por día (en horas)
demand = {"Lunes": 136, "Martes": 104, "Miércoles": 120, "Jueves": 152, "Viernes": 112, "Sábado": 128, "Domingo": 88}

# Definir variables de decisión
X_FT = {d: LpVariable(f"FT_{d}", 0, None, cat="Integer") for d in days}  # Full-time employees
X_PT = {d: LpVariable(f"PT_{d}", 0, None, cat="Integer") for d in days}  # Part-time employees

# Costos por hora y horas por semana
cost_FT = 15  # USD por hora
cost_PT = 10  # USD por hora
hours_FT = 8 * 5  # 40 horas a la semana por empleado FT
hours_PT = 4 * 5  # 20 horas a la semana por empleado PT

# Definir el problema de optimización
prob = LpProblem("Minimizar_Costo_Empleados", LpMinimize)

# Función objetivo: Minimizar costos semanales reales
prob += lpSum([cost_FT * hours_FT * X_FT[d] + cost_PT * hours_PT * X_PT[d] for d in days])

# Restricciones de demanda diaria: Asegurar que se cubran las horas necesarias cada día
for i, d in enumerate(days):
    prob += (lpSum(8 * X_FT[days[j % num_days]] + 4 * X_PT[days[j % num_days]] for j in range(i, i + 5)) >= demand[d])

# Restricción del 25% de empleados medio tiempo (en términos de horas semanales)
total_hours_week = sum(demand.values())
prob += lpSum(hours_PT * X_PT[d] for d in days) <= 0.25 * total_hours_week

# Resolver el problema
prob.solve()

# Imprimir resultados
for d in days:
    print(f"Día {d}: FT = {X_FT[d].varValue}, PT = {X_PT[d].varValue}")
print("Costo total mínimo =", prob.objective.value())


Día Lunes: FT = 7.0, PT = 1.0
Día Martes: FT = 0.0, PT = 0.0
Día Miércoles: FT = 3.0, PT = 0.0
Día Jueves: FT = 0.0, PT = 0.0
Día Viernes: FT = 3.0, PT = 7.0
Día Sábado: FT = 4.0, PT = 1.0
Día Domingo: FT = 1.0, PT = 0.0
Costo total mínimo = 12600.0


#### **Conclusión y Recomendaciones para la Post Office**  

El modelo de optimización ha determinado la **asignación óptima de trabajadores** a lo largo de la semana para minimizar los costos operativos mientras se cumplen los requisitos de demanda laboral.  

---

#### **Distribución Óptima del Personal**  

| Día       | Empleados de Tiempo Completo (FT) | Empleados de Medio Tiempo (PT) |
|-----------|---------------------------------|-----------------------------|
| **Lunes**  | 7   | 1  |
| **Martes**  | 0   | 0  |
| **Miércoles**  | 3   | 0  |
| **Jueves**  | 0   | 0  |
| **Viernes**  | 3   | 7  |
| **Sábado**  | 4   | 1  |
| **Domingo**  | 1   | 0  |

✅ **Costo total mínimo alcanzado:** **$12,600**  
✅ **Se minimizó la cantidad de empleados sin afectar la operación**  

---

#### **Puntos Clave y Recomendaciones**  

1. **No se requiere personal los martes ni los jueves**  
   - El modelo confirma que en estos días **no se necesita personal** para cumplir con la carga laboral mínima.  
   - Esto sugiere que en estos días la demanda de trabajo es baja o inexistente.  

2. **La mayor cantidad de empleados es requerida los viernes**  
   - **10 trabajadores en total (3 FT y 7 PT)** son necesarios para cubrir las necesidades del día.  
   - Esto sugiere que el viernes es un día de alta demanda en la oficina postal.  

3. **Uso eficiente de empleados de tiempo parcial (PT)**  
   - Se asignan solo en los días **viernes (7 PT), lunes (1 PT) y sábado (1 PT)**.  
   - Esto permite reducir costos al no contratar más empleados de los necesarios.  

4. **El modelo encontró la combinación más barata de personal bajo las restricciones**  
   - Se optimizó el número de empleados para cumplir con la carga laboral **al menor costo posible ($12,600)**.  
   - Se evitaron contrataciones innecesarias, maximizando la eficiencia del presupuesto.  

---

#### **Recomendación Final**  
La **Post Office** debería seguir esta asignación de empleados para operar con el **mínimo costo posible**, asegurando que los días clave con mayor demanda (lunes, miércoles, viernes y sábado) estén cubiertos de manera óptima.  

Además, puede analizar **por qué los martes y jueves no se requieren empleados** y, si es viable, reorganizar algunas tareas para balancear mejor la carga de trabajo.  

**Conclusión:** Se logró minimizar costos sin afectar el servicio, asegurando una operación eficiente y rentable.



### 📌 Solución en Excel  
Este punto puedes verlo resuelto en Excel a través de Solver en el siguiente enlace en la hoja llamada "Ejercicio 2 B.":  
[Haz clic aquí para verlo](https://docs.google.com/spreadsheets/d/1aVTQIgdwYh91xQ1oS19cvkOjbDoZUB89/edit?usp=sharing&ouid=102065898380876538162&rtpof=true&sd=true)

##**3. CIUDAD ERSTVILLE Y SUS IMPUESTOS**
La ciudad de Erstville enfrenta un grave recorte de presupuesto.

Buscando una solución a largo plazo para mejorar la base tributaria, el consejo de la ciudad propone la demolición de un área de viviendas dentro de la ciudad, y su reemplazo con un moderno desarrollo.

El proyecto implica dos fases: (1) demolición de casas populares para obtener el terreno para el nuevo desarrollo, y (2) construcción del nuevo desarrollo.
A continuación, un resumen de la situación:

Se pueden demoler 300 casas populares. Cada casa ocupa un lote de .25 acres. El costo de demoler una casa es de $2000.

Los tamaños de los lotes para construir casas unifamiliares, dobles, triples y cuádruples, son de .18, .28, .4 y .5 acres, respectivamente.
Las calles, los espacios abiertos y el área para la instalación de servicios, ocupan 15% del área disponible.

En el nuevo desarrollo, las unidades triples y cuádruples ocupan por lo menos 25% del total del espacio construido. El número de unidades sencillas deben ser al menos 20% de todas las unidades, y las unidades dobles deben ocupar un mínimo de 10% del total del espacio construido.

El impuesto por unidad aplicado a las unidades sencillas, dobles, triples y cuádruples es de 1000, 1900, 2700 y 3400 USD respectivamente.

El costo de construcción por unidad de las casas sencillas, dobles, triples y cuádruples es de 50,000, 70,000, 130,000 y 160,000 USD respectivamente.

El financiamiento a través de un banco local está limitado a 15 millones de USD.

¿Cuántas unidades de cada tipo se deben construir para maximizar la recaudación de impuestos?

In [ ]:
# Definir el problema de optimización
problem = LpProblem("Maximizar_Impuestos", LpMaximize)

# Variables de decisión
x1 = LpVariable("Unidades_Sencillas", 0, None, cat="Integer")
x2 = LpVariable("Unidades_Dobles", 0, None, cat="Integer")
x3 = LpVariable("Unidades_Triples", 0, None, cat="Integer")
x4 = LpVariable("Unidades_Cuadruples", 0, None, cat="Integer")
D = LpVariable("Casas_Demolidas", 0, 300, cat="Integer")


# Función objetivo: Maximizar los impuestos recaudados
problem += 1000*x1 + 1900*x2 + 2700*x3 + 3400*x4

# Restricción de área disponible (85% del terreno liberado por la demolición)
problem += (0.18*x1 + 0.28*x2 + 0.40*x3 + 0.50*x4) <= (0.85 * (D * 0.25))

# Restricción de presupuesto total (incluyendo demolición)
problem += (50000*x1 + 70000*x2 + 130000*x3 + 160000*x4 + 2000*D) <= 15000000

# Restricción basada en área: Triples y cuádruples ≥ 25% del espacio total
problem += (0.40*x3 + 0.50*x4) >= 0.25 * (0.18*x1 + 0.28*x2 + 0.40*x3 + 0.50*x4)

# Restricción basada en área: Dobles ≥ 10% del espacio total
problem += (0.28*x2) >= 0.10 * (0.18*x1 + 0.28*x2 + 0.40*x3 + 0.50*x4)

# Restricción basada en área: Sencillas ≥ 20% del total de casas construidas
problem += (x1) >= 0.20 * (x1 + x2 + x3 + x4)

# Resolver el problema
problem.solve()

# Mostrar resultados
print("Solución óptima:")
print(f"Casas a demoler: {D.varValue}")
print(f"Unidades sencillas: {x1.varValue}")
print(f"Unidades dobles: {x2.varValue}")
print(f"Unidades triples: {x3.varValue}")
print(f"Unidades cuádruples: {x4.varValue}")
print(f"Recaudación máxima de impuestos: $ {problem.objective.value()}")


Solución óptima:
Casas a demoler: 252.0
Unidades sencillas: 37.0
Unidades dobles: 119.0
Unidades triples: 1.0
Unidades cuádruples: 26.0
Recaudación máxima de impuestos: $ 354200.0


In [ ]:
print("\nResumen del uso del área construida:")
print(f"Área total construida: {0.18*x1.varValue + 0.28*x2.varValue + 0.40*x3.varValue + 0.50*x4.varValue:.2f} acres")
print(f"Área de unidades sencillas: {0.18*x1.varValue:.2f} acres")
print(f"Área de unidades dobles: {0.28*x2.varValue:.2f} acres")
print(f"Área de unidades triples: {0.40*x3.varValue:.2f} acres")
print(f"Área de unidades cuádruples: {0.50*x4.varValue:.2f} acres")


Resumen del uso del área construida:
Área total construida: 53.38 acres
Área de unidades sencillas: 6.66 acres
Área de unidades dobles: 33.32 acres
Área de unidades triples: 0.40 acres
Área de unidades cuádruples: 13.00 acres


### **Conclusión del modelo de optimización para la ciudad de Erstville**  

Con base en los resultados obtenidos, la estrategia óptima para **maximizar la recaudación de impuestos** en Erstville implica:  

- **Demoler 252 casas populares**  
- Construir **37** casas **unifamiliares**  
- Construir **119** casas **dobles**  
- Construir **1** casa **triple**  
- Construir **26** casas **cuádruples**  

Esta planificación permite alcanzar una **recaudación máxima de impuestos de $354,200**, cumpliendo con todas las restricciones del problema, incluyendo:  
✅ **Disponibilidad de área para construcción**  
✅ **Proporción mínima de unidades sencillas, dobles, triples y cuádruples**  
✅ **Límite de financiamiento disponible**  

### **Conclusiones clave para la planeación del desarrollo**  

1. **Se requiere la demolición de 252 casas populares**  
   - No es necesario demoler todas las 300 casas disponibles.  
   - La optimización sugiere que con 252 casas demolidas se logra la mejor distribución de espacio para maximizar la recaudación.  

2. **Enfoque en casas dobles**  
   - Se construyen **119 casas dobles**, representando la mayor cantidad de unidades en el desarrollo.  
   - Esto indica que este tipo de vivienda es la más eficiente en términos de recaudación y uso del espacio.  

3. **Mínima construcción de casas triples**  
   - Solo **1 unidad triple** es necesaria para cumplir con las restricciones y maximizar la recaudación.  
   - Esto sugiere que su aporte a la recaudación es bajo en comparación con otros tipos de unidades.  

4. **Uso eficiente del financiamiento**  
   - El presupuesto máximo disponible de **$15 millones** se utiliza estratégicamente dentro del límite permitido.  
   - La solución no requiere financiamiento adicional para lograr la mayor recaudación posible.  

5. **Cumplimiento de restricciones de zonificación y espacio**  
   - Se respeta la condición de que las **unidades triples y cuádruples ocupen al menos el 25% del espacio total**.  
   - Se cumple con la restricción de que **las unidades sencillas sean al menos el 20% del total**.  
   - También se respeta la regla de que **las unidades dobles ocupen al menos el 10% del área construida**.  

### **Recomendación para la ciudad de Erstville**  
Para obtener la **máxima recaudación de impuestos** dentro de las restricciones de financiamiento y espacio, la ciudad debe priorizar la construcción de **casas dobles**, seguidas de **casas sencillas y cuádruples**, manteniendo solo una mínima cantidad de unidades triples.  

Esta estrategia permitirá **optimizar la base tributaria de la ciudad**, asegurando un desarrollo sostenible y eficiente en términos de espacio y presupuesto.  


### 📌 Solución en Excel  
Este punto puedes verlo resuelto en Excel a través de Solver en el siguiente enlace en la hoja llamada "Ejercicio 3":  
[Haz clic aquí para verlo](https://docs.google.com/spreadsheets/d/1aVTQIgdwYh91xQ1oS19cvkOjbDoZUB89/edit?usp=sharing&ouid=102065898380876538162&rtpof=true&sd=true)

##**4. PLANEACION JUGUITOS SABROSOS**

Juguitos sabrosos elabora dos tipos de jugo de naranja Premium y regular.
Para elaborar estos productos utiliza 2 tipos de naranjas oromiel y normales.

Las naranjas oromiel tienen un grado de azúcar de 6 y las normales de 3.

El juego de naranja Premium debe contener en promedio 5 grados de azúcar mientras que el jugo regular debe tener 4.

Cada litro de jugo de naranja Premium se vende a 4000 mientras que el regular se vende a 2000.

Por otra parte, un kilo de naranjas de tipo oromiel se consigue a 1500 por kilo mientras que las normales cuestan 1000 por kg. (Asuma que cada kilo de naranjas produce un litro de jugo independientemente del tipo).

La demanda de jugos es de 30 litros de premium y 20 litros de regular


In [ ]:
# Definir el problema
model = LpProblem("Maximizar_Ganancia", LpMaximize)

# Definir variables de decisión (Kg de cada tipo de naranja)
X_op = LpVariable("X_op", lowBound=0)  # Oromiel para Premium
X_np = LpVariable("X_np", lowBound=0)  # Normal para Premium
X_or = LpVariable("X_or", lowBound=0)  # Oromiel para Regular
X_nr = LpVariable("X_nr", lowBound=0)  # Normal para Regular

# Función objetivo (Maximizar ganancia)
model += (4000 * 30 + 2000 * 20) - (1500 * X_op + 1000 * X_np + 1500 * X_or + 1000 * X_nr)

# Restricciones de producción (demanda)
model += X_op + X_np == 30  # Total Premium
model += X_or + X_nr == 20  # Total Regular

# Restricciones de azúcar promedio
model += (6 * X_op + 3 * X_np) / 30 == 5  # Azúcar Premium
model += (6 * X_or + 3 * X_nr) / 20 == 4  # Azúcar Regular

# Resolver el problema
model.solve()

# Imprimir resultados
print("Resultados:")
print(f"Kg de naranjas Oromiel para Premium: {X_op.varValue}")
print(f"Kg de naranjas Normales para Premium: {X_np.varValue}")
print(f"Kg de naranjas Oromiel para Regular: {X_or.varValue}")
print(f"Kg de naranjas Normales para Regular: {X_nr.varValue}")
print(f"Ganancia máxima: $ {model.objective.value()}")


Resultados:
Kg de naranjas Oromiel para Premium: 20.0
Kg de naranjas Normales para Premium: 10.0
Kg de naranjas Oromiel para Regular: 6.6666667
Kg de naranjas Normales para Regular: 13.333333
Ganancia máxima: $ 96666.66695


Podríamos darle al dueño de **Juguitos Sabrosos** las siguientes recomendaciones basadas en los resultados del modelo de optimización:  

### **Recomendaciones para la planeación de producción**  

1. **Distribución de insumos óptima:**  
   - Para producir los **30 litros de jugo Premium**, debe usar:  
     - **20 kg de naranjas Oromiel**  
     - **10 kg de naranjas Normales**  
   - Para producir los **20 litros de jugo Regular**, debe usar:  
     - **6.67 kg de naranjas Oromiel**  
     - **13.33 kg de naranjas Normales**  

2. **Cumplimiento de calidad:**  
   - Esta combinación de insumos garantiza que el jugo Premium tenga un **promedio de 5 grados de azúcar** y el Regular **4 grados**, cumpliendo con los estándares de calidad.  

3. **Ganancia máxima esperada:**  
   - Con esta estrategia, la empresa puede lograr una **ganancia máxima de $96,666.67**, considerando costos de materia prima y precio de venta.  

4. **Consideraciones adicionales:**  
   - Si el precio de las naranjas cambia o hay variaciones en la demanda, podría ser necesario **recalcular la estrategia** para maximizar la ganancia.  
   - Si la empresa desea aumentar la producción o explorar nuevas combinaciones de insumos, se pueden realizar más análisis de optimización.  

### **Conclusión**  
La empresa puede seguir este plan de producción para maximizar sus ganancias mientras cumple con los requerimientos de calidad de sus productos.

### 📌 Solución en Excel  
Este punto puedes verlo resuelto en Excel a través de Solver en el siguiente enlace en la hoja llamada "Ejercicio 4":  
[Haz clic aquí para verlo](https://docs.google.com/spreadsheets/d/1aVTQIgdwYh91xQ1oS19cvkOjbDoZUB89/edit?usp=sharing&ouid=102065898380876538162&rtpof=true&sd=true)